pkg/backend/httpstate/client/api.go

```diff
@@ -252,3 +286,36 @@ func pulumiRESTCall(ctx context.Context, diag diag.Sink, cloudAPI, method, path
 
 	return nil
 }
+
+// readBody reads the contents of an http.Response into a byte array, returning an error if one occurred while in the
+// process of doing so. readBody uses the Content-Encoding of the response to pick the correct reader to use.
+func readBody(resp *http.Response) ([]byte, error) {
+	contentEncoding, ok := resp.Header["Content-Encoding"]
+	body := resp.Body
+	defer contract.IgnoreClose(body)
+	if !ok {
+		// no header -> no additional encoding
+		return ioutil.ReadAll(resp.Body)
+	}
+
+	if len(contentEncoding) > 1 {
+		// we only know how to deal with gzip. We can't handle additional encodings layered on top of it.
+		return nil, errors.Errorf("can't read body: can't handle content encodings %v", contentEncoding)
+	}
+
+	switch contentEncoding[0] {
+	case "x-gzip":
+		// The HTTP/1.1 spec recommends we treat x-gzip as an alias of gzip
+		fallthrough
+	case "gzip":
+		logging.V(7).Infoln("decompressing gzipped response from service")
+		reader, err := gzip.NewReader(resp.Body)
+		if err != nil {
+			return nil, errors.Wrap(err, "can't read body")
+		}
+
+		return ioutil.ReadAll(reader)
+	default:
+		return nil, errors.Errorf("can't read body: unrecognized encoding %s", contentEncoding[0])
```

Comment:
Is "can't read body" redundant? As the error is being returned from a function called `readBody`? IMHO  "unrecognized encoding XXX" is a perfectly acceptable error.